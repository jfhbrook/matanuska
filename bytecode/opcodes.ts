export enum OpCode {
  // Constants and values
  Constant,
  Nil,
  True,
  False,

  // !
  Pop,

  // Operators
  //
  // TODO: BASIC is typed - we can check the types in the compiler and use
  // typed op codes. However, this requires tracking the state of the stack
  // in the compiler. I'm keeping the opcodes untyped for now, just to get
  // something working, but would like to switch to typed opcodes later.

  // Comparison operators
  Eq,
  Gt,
  Ge,
  Lt,
  Le,
  Ne,

  // Operators
  Not,
  Add,
  Sub,
  Mul,
  Div,
  Neg,

  // Commands
  Print,

  // Jumping
  Jump,
  JumpIfFalse,
  Loop,

  Return,
}
