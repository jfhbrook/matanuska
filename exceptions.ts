import { ExitCode, ExitCoded } from './exit';
import { ErrorCode } from './errors';
import { Formattable, Formatter, FormatValue } from './format';
import { Traceable, Traceback } from './traceback';

/**
 * The base class for all Exceptions, including fatal exceptions.
 */
export class BaseException implements Traceable, Formattable {
  /**
   * @param message The message for the exception.
   * @param traceback The traceback for the exception.
   */
  constructor(
    public message: FormatValue,
    public traceback: Traceback | null,
  ) {
    // TODO: Exceptions in Python store all args the constructor is called
    // with, not just a message.
    // TODO: Implement notes a la Python?
    // See: https://docs.python.org/3/tutorial/errors.html#enriching-exceptions-with-notes
  }

  format(formatter: Formatter): string {
    return formatter.formatBaseException(this);
  }
}

/**
 * The base class for all built-in, non-fatal exceptions. User
 * exceptions should be derived from this class.
 */
export class Exception extends BaseException {}

//
// Non-fatal errors.
//

/**
 * An exception raised by an assertion.
 */
export class AssertionError extends Exception {
  format(formatter: Formatter): string {
    return formatter.formatAssertionError(this);
  }
}

/**
 * An exception raised on errors in the runtime.
 */
export class RuntimeError extends Exception {}

/**
 * An exception raised when functionality is not implemented. Extends
 * RuntimeError.
 */
export class NotImplementedError extends RuntimeError {}

//
// Warnings.
//

/**
 * The base class for all warnings. Warnings are non-fatal and can not be
 * caught.
 */
export class BaseWarning extends BaseException {
  format(formatter: Formatter): string {
    return formatter.formatBaseWarning(this);
  }
}

/**
 * The base class for all user warnings.
 */
export class Warning extends BaseWarning {}

/**
 * A warning for deprecated functionality.
 */
export class DeprecationWarning extends Warning {}

//
// OS and file IO errors.
//

/**
 * An operating system level error.
 */
export class OsError extends Exception implements ExitCoded {
  public readonly exitCode: ExitCode;

  /**
   * @param message The message for the exception.
   * @param code The Node.JS system error code.
   * @param exitCode The exit code for the exception. If null, a default exit
   *        code will be set based on the error code.
   * @param traceback The traceback for the exception.
   */
  constructor(
    public readonly message: FormatValue,
    public readonly code: ErrorCode | string,
    exitCode: ExitCode | null,
    public readonly traceback: Traceback | null,
  ) {
    super(message, traceback);

    if (exitCode) {
      this.exitCode = exitCode;
    } else {
      switch (code) {
        // NOTE: For FileErrors, ErrorCode.Access maps to ExitCode.NoInput
        // for read operations and ExitCode.CantCreate for write operations.
        case ErrorCode.Access:
          this.exitCode = ExitCode.OsError;
          break;
        case ErrorCode.AddressInUse:
          // There are arguments for this being Unavailable, OsError or
          // Software.
          this.exitCode = ExitCode.Protocol;
          break;
        case ErrorCode.ConnectionRefused:
        case ErrorCode.ConnectionReset:
          this.exitCode = ExitCode.Unavailable;
          break;
        case ErrorCode.Exists:
          this.exitCode = ExitCode.CantCreate;
          break;
        case ErrorCode.IsDirectory:
          this.exitCode = ExitCode.IoError;
          break;
        case ErrorCode.MaxFileDescriptors:
          this.exitCode = ExitCode.OsError;
          break;
        case ErrorCode.NoEntity:
          this.exitCode = ExitCode.NoInput;
          break;
        case ErrorCode.NotADirectory:
          this.exitCode = ExitCode.NoInput;
          break;
        case ErrorCode.NotEmpty:
          // TODO: There's an argument for this being an IoError.
          this.exitCode = ExitCode.OsError;
          break;
        case ErrorCode.DnsNotFound:
          this.exitCode = ExitCode.NoHost;
          break;
        case ErrorCode.NotPermitted:
          // NOTE: Assuming this will not be seen for user filesystem
          // permissions.
          this.exitCode = ExitCode.NoPermission;
          break;
        case ErrorCode.BrokenPipe:
          this.exitCode = ExitCode.OsError;
          break;
        case ErrorCode.TimedOut:
          // NOTE: Unavailable is appropriate for network timeouts. For other
          // kinds of timeout, there may be better exit codes.
          this.exitCode = ExitCode.Unavailable;
          break;
        default:
          this.exitCode = ExitCode.OsError;
      }
    }
  }

  format(formatter: Formatter): string {
    return formatter.formatOsError(this);
  }
}

/**
 * The path(s) being used by an operations which raised a FileError.
 */
export type FileErrorPaths = [string] | [string, string];

/**
 * An error raised when a known file operation fails. Includes the target
 * filename(s) and a more specific exit code.
 */
export class FileError extends OsError {
  /**
   * @param message The message for the exception.
   * @param code The Node.JS system error code.
   * @param exitCode The exit code for the exception. If null, a default exit
   *        code will be set based on the error code.
   * @param paths The target path(s) for the operation.
   * @param traceback The traceback for the exception.
   */
  constructor(
    public message: FormatValue,
    public code: ErrorCode | string,
    exitCode: ExitCode | null,
    public paths: FileErrorPaths,
    public traceback: Traceback | null,
  ) {
    super(message, code, exitCode, traceback);
  }

  /**
   * Naively create a FileError from a Node.js error.
   *
   * @param message The message for the error. If null, taken directly from
   * the Node.JS error.
   * @param err The Node.JS error.
   * @param traceback The traceback for the error.
   */
  static fromError(
    message: FormatValue | null,
    err: NodeJS.ErrnoException,
    traceback: Traceback | null,
  ) {
    return new FileError(
      message || err.message,
      err.code,
      null,
      [err.path],
      traceback,
    );
  }

  /**
   * Create a FileError from a Node.js error for a read operation.
   *
   * @param message The message for the error. If null, taken directly from
   * the Node.JS error.
   * @param err The Node.JS error.
   * @param traceback The traceback for the error.
   */
  static fromReadError(
    message: FormatValue | null,
    err: NodeJS.ErrnoException,
    traceback: Traceback | null,
  ) {
    const code = err.code;
    let exitCode: ExitCode | null = null;

    if (code == ErrorCode.Access) {
      exitCode = ExitCode.NoInput;
    }

    return new FileError(
      message || err.message,
      err.code,
      exitCode,
      [err.path],
      traceback,
    );
  }

  /**
   * Create a FileError from a Node.js error for a write operation.
   *
   * @param message The message for the error. If null, taken directly from
   * the Node.JS error.
   * @param err The Node.JS error.
   * @param traceback The traceback for the error.
   */
  static fromWriteError(
    message: FormatValue | null,
    err: NodeJS.ErrnoException,
    traceback: Traceback | null,
  ) {
    const code = err.code;
    let exitCode: ExitCode | null = null;

    if (code == ErrorCode.Access) {
      exitCode = ExitCode.CantCreate;
    }

    return new FileError(
      message || err.message,
      err.code,
      exitCode,
      [err.path],
      traceback,
    );
  }

  format(formatter: Formatter): string {
    return formatter.formatFileError(this);
  }
}

//
// Syntax and parse errors and warnings.
//
// TODO: JavaScript and Python stop after a single error. However, Matanuska
// will attempt to complete the parse and flag multiple errors, more like C.
// This leads to some questions in their interfaces:
//
// 1. Individual syntax errors and warnings inherit from BaseException and
//    BaseWarning respectively. However, they don't have tracebacks - rather,
//    they have source locations. This suggests, along with the fact that they
//    are never thrown, that it may be inappropriate for them to be
//    Traceable - or even be called errors/warnings. Note that
//    the reporting in Python for source locations is very similar to
//    tracebacks, just without the word "Traceback".
// 2. ParseErrors and ParseWarnings also do not take tracebacks. This is
//    because parsing happens before runtime execution can occur. It may,
//    however, make sense to revisit this assumption.
// 2. ParseErrors and ParseWarnings have blank messages. This is acceptable
//    because the reporting will be based on the syntax errors/warnings, not
//    the top-level parse error. Should these classes have a message
//    getter that calls a formatter? Should they have a sensible default
//    message? This is an open question.
//

/**
 * The location referenced by a syntax error or warning.
 */
export interface SourceLocation {
  /**
   * The filename for the source code. This may be <interactive> for
   * interactive code, or <untitled> if the edited code has not been saved.
   */
  filename: string;

  /**
   * Whether or not the source location is in a line.
   */
  isLine: boolean;

  /**
   * The line number in the source code. When not in a line, this is the
   * previous line number.
   */
  lineNo: number;

  /**
   * The column offset at which the location starts.
   */
  readonly offsetStart: number;

  /**
   * The column offset at which the location ends.
   */
  readonly offsetEnd: number;



  /**
   * The referenced source code. Will typically be a line or interactive
   * command.
   */
  source: string;

  // endLineNo: number | null,
  // endOffset: number | null
}

/**
 * An individual syntax error. Typically there will be multiple syntax errors
 * or warnings encapsulated in a ParseError. They are not intended to be
 * thrown.
 */
export class SyntaxError extends BaseException implements SourceLocation {
  /**
   * @param message The message for the exception.
   * @param filename The filename for the source code.
   * @param lineNo The line number in the source code.
   * @param offset The column offset in the source code.
   * @param source The referenced source code.
   * @param traceback The traceback for the exception.
   */
  constructor(
    message: FormatValue,
    public filename: string,
    public lineNo: number,
    public offsetStart: number,
    public offsetEnd: number,
    public source: string,
  ) {
    super(message, null);
  }

  format(formatter: Formatter): string {
    return formatter.formatSyntaxError(this);
  }
}

/**
 * A parse error. May contain one or more syntax errors or warnings.
 */
export class ParseError extends Exception implements ExitCoded {
  public exitCode = ExitCode.Software;

  /**
   * @param message The message for the exception.
   * @param errors A collection of syntax errors and warnings.
   * @param traceback The traceback for the exception.
   */
  constructor(public errors: Array<SyntaxError | SyntaxWarning>) {
    super('', null);
  }

  format(formatter: Formatter): string {
    return formatter.formatParseError(this);
  }
}

/**
 * A syntax warning. These warnings don't cause an unsuccessful parse, but are
 * likely undesirable. Typically there will be multiple syntax warnings
 * encapsulated in a ParseWarning. They are not intended to be logged
 * directly.
 */
export class SyntaxWarning extends BaseWarning implements SourceLocation {
  /**
   * @param message The message for the exception.
   * @param filename The filename for the source code.
   * @param lineNo The line number in the source code.
   * @param offset The column offset in the source code.
   * @param source The referenced source code.
   * @param traceback The traceback for the exception.
   */
  constructor(
    message: FormatValue,
    public filename: string,
    public lineNo: number,
    public offsetStart: number,
    public offsetEnd: number,
    public source: string,
  ) {
    super(message, null);
  }

  format(formatter: Formatter): string {
    return formatter.formatSyntaxWarning(this);
  }
}

/**
 * A parse warning. This is analogous to a syntax error, but contains only
 * warnings.
 */
export class ParseWarning extends Exception {
  /**
   * @param message The message for the exception.
   * @param warnings A collection of syntax warnings.
   * @param traceback The traceback for the exception.
   */
  constructor(public warnings: Array<SyntaxWarning>) {
    super('', null);
  }

  format(formatter: Formatter): string {
    return formatter.formatParseWarning(this);
  }
}
