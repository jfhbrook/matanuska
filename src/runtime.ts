import { Host } from './host';
import { Errors } from './errors';
import { Line } from './internal';

// Validates the program prior to running. The editor can do a lot syntax
// error checking during edit time, but anything involving context - for
// example, for loops being closed - happens here.
export function preRun(program: Line[]): void {
}

export interface RuntimeOptions {
  host: Host;
  errors: Errors;
}

export class Runtime {
  private readonly host: Host;
  private readonly errors: Errors;

  constructor({ host, errors }: RuntimeOptions) {
    this.host = host;
    this.errors = errors;
  }
}
