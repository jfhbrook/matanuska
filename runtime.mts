import * as assert from 'node:assert';

import { startTraceExec, traceExec } from './debug.mjs';

import {
  BaseException,
  NameError,
  NotImplementedError,
} from './exceptions.mjs';
import { Exit } from './exit.mjs';
import { RuntimeFault } from './faults.mjs';
import { Host } from './host.mjs';
import { Stack } from './stack.mjs';
import { Traceback } from './traceback.mjs';
import { Value, nil, Nil } from './value/index.mjs';
import { falsey } from './value/truthiness.mjs';

import { Byte } from './bytecode/byte.mjs';
import { Chunk } from './bytecode/chunk.mjs';
import { OpCode } from './bytecode/opcodes.mjs';
import { Short, bytesToShort } from './bytecode/short.mjs';

import * as op from './operations.mjs';

export type Globals = Record<string, Value>;

export class Runtime {
  public stack: Stack<Value>;
  public pc: number = -1;
  public chunk: Chunk = new Chunk();
  public globals: Globals = {};

  constructor(private host: Host) {
    this.stack = new Stack();
  }

  public reset(): void {
    this.stack = new Stack();
    this.chunk = new Chunk();
    this.pc = 0;
  }

  public interpret(chunk: Chunk): Value | null {
    this.chunk = chunk;
    this.pc = 0;
    return this.run();
  }

  // TODO: Using templates can help decrease boilerplate while increasing the
  // amount of inlining. But if I'm going down that road, I might want to
  // consider porting this to C++ anyway.

  private readByte(): Byte {
    const byte = this.chunk.code[this.pc];
    this.pc++;
    return byte;
  }

  private readCode(): OpCode {
    return this.readByte();
  }

  private readConstant(): Value {
    return this.chunk.constants[this.readByte()];
  }

  private readShort(): Short {
    return bytesToShort([this.readByte(), this.readByte()]);
  }

  private readString(): string {
    const value = this.readConstant();
    assert.equal(typeof value, 'string', 'Value is string');
    return value as string;
  }

  private createTraceback(): Traceback | null {
    return new Traceback(
      null,
      this.chunk.filename,
      this.chunk.lines[this.pc - 1],
    );
  }

  private run(): Value | null {
    let a: Value | null = null;
    let b: Value | null = null;

    startTraceExec();

    try {
      while (true) {
        traceExec(this);
        const instruction = this.readCode();

        switch (instruction) {
          case OpCode.Constant:
            this.stack.push(this.readConstant());
            break;
          case OpCode.Nil:
            this.stack.push(nil);
            break;
          case OpCode.True:
            this.stack.push(true);
            break;
          case OpCode.False:
            this.stack.push(false);
            break;
          case OpCode.Pop:
            this.stack.pop();
            break;
          case OpCode.GetGlobal:
            a = this.readString();
            b = this.globals[a];
            if (typeof b === 'undefined') {
              throw new NameError(`Variable ${a} is undefined`);
            }
            this.stack.push(b);
            break;
          case OpCode.DefineGlobal:
            a = this.readString();
            // NOTE: Pops afterwards for garbage collection reasons
            b = this.stack.peek();
            if (typeof this.globals[a] !== 'undefined') {
              throw new NameError(`Cannot define variable ${a} twice`);
            }
            this.globals[a] = b as Value;
            this.stack.pop();
            break;
          case OpCode.SetGlobal:
            a = this.readString();
            // NOTE: Pops afterwards for garbage collection reasons
            b = this.stack.peek();
            if (typeof this.globals[a] === 'undefined') {
              throw new NameError(`Cannot assign to undefined variable ${a}`);
            }
            this.globals[a] = b as Value;
            // TODO: This is missing from my clox implementation. That's a
            // bug, right?
            this.stack.pop();
            break;
          case OpCode.Eq:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(op.eq(a, b));
            break;
          case OpCode.Gt:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(op.gt(a, b));
            break;
          case OpCode.Ge:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(op.ge(a, b));
            break;
          case OpCode.Lt:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(op.lt(a, b));
            break;
          case OpCode.Le:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(op.le(a, b));
            break;
          case OpCode.Ne:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(op.ne(a, b));
            break;
          case OpCode.Not:
            a = this.stack.pop();
            this.stack.push(op.not(a));
            break;
          case OpCode.Add:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(op.add(a, b));
            break;
          case OpCode.Sub:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(op.sub(a, b));
            break;
          case OpCode.Mul:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(op.mul(a, b));
            break;
          case OpCode.Div:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(op.div(a, b));
            break;
          case OpCode.Neg:
            a = this.stack.pop();
            this.stack.push(op.neg(a));
            break;
          case OpCode.Print:
            this.host.writeLine(this.stack.pop());
            break;
          case OpCode.Exit:
            const value = this.stack.pop();
            let exitCode: number;
            if (typeof value === 'number') {
              exitCode = Math.floor(value);
            } else if (value instanceof Nil) {
              exitCode = 0;
            } else if (value) {
              exitCode = 1;
            } else {
              exitCode = 0;
            }
            this.host.exit(exitCode);
            return null;
          case OpCode.Jump:
            // Note, readShort increments the pc. If we didn't assign before,
            // we would need to add extra to skip over those bytes!
            a = this.readShort();
            this.pc += a;
            break;
          case OpCode.JumpIfFalse:
            a = this.readShort();
            b = this.stack.peek();

            if (falsey(b!)) {
              this.pc += a;
            }
            break;
          case OpCode.Loop:
            this.notImplemented('Loop');
            break;
          case OpCode.Return:
            a = this.stack.pop();
            // TODO: Clean up the current frame, and only return if we're
            // done with the main program.
            return a;
          default:
            assert.ok(
              this.pc < this.chunk.code.length,
              'Program counter out of bounds',
            );
            this.notImplemented(`Unknown opcode: ${instruction}`);
        }
      }
    } catch (err) {
      if (err instanceof Exit) {
        throw err;
      }

      if (!(err instanceof BaseException)) {
        err = RuntimeFault.fromError(err);
      }

      err.traceback = this.createTraceback();
      throw err;
    }
  }

  private notImplemented(message: string): Value {
    throw new NotImplementedError(message);
  }
}
