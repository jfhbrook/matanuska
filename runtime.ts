import { spanSync } from './trace';

import { NotImplementedError } from './exceptions';
import { Host } from './host';
import { Stack } from './stack';
import { Value, nil } from './value';

import { Chunk } from './bytecode/chunk';
import { OpCode } from './bytecode/opcodes';

export class Runtime {
  public stack: Stack;
  private pc: number = -1;
  private chunk: Chunk | null = null;

  constructor(private host: Host) {
    this.stack = new Stack();
  }

  interpret(chunk: Chunk): Value {
    this.chunk = chunk;
    this.pc = 0;
    return this.run();
  }

  // TODO: Using templates can help decrease boilerplate while increasing the
  // amount of inlining. But if I'm going down that road, I might want to
  // consider porting this to C++ anyway.

  // NOTE: I'm calling this readByte to be consistent with clox, but in truth
  // we're not reading bytes at all. See chunk.ts for more details.
  //
  // In the future, I might rename this.
  readByte(): OpCode {
    const code = this.chunk.code[this.pc];
    this.pc++;
    return code;
  }

  readConstant(): Value {
    return this.chunk.constants[this.readByte()];
  }

  run(): Value {
    return spanSync('run', () => {
      let a: any = null;
      let b: any = null;
      let rv: Value | null = null;

      while (true) {
        const instruction = this.readByte();

        switch (instruction) {
          case OpCode.Constant:
            this.stack.push(this.readConstant());
            break;
          case OpCode.Nil:
            this.stack.push(nil);
            break;
          case OpCode.True:
            this.stack.push(true);
            break;
          case OpCode.False:
            this.stack.push(false);
            break;
          case OpCode.Pop:
            this.stack.pop();
            break;
          //
          // TODO: These operators don't do any type checking right now. At a
          // minimum I need to check what the types of a/b are and switch
          // accordingly. But I may also want separate instructions for
          // different types. I may also want to have particular instructions
          // for type conversions.
          //
          // TODO: Define my own semantics for equality.
          //
          case OpCode.Eq:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(a === b);
            break;
          case OpCode.Gt:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(a > b);
            break;
          case OpCode.Ge:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(a >= b);
            break;
          case OpCode.Lt:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(a < b);
            break;
          case OpCode.Le:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(a <= b);
            break;
          case OpCode.Ne:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(a !== b);
            break;
          case OpCode.Not:
            a = this.stack.pop();
            this.stack.push(!a);
            break;
          case OpCode.Add:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(a + b);
            break;
          case OpCode.Sub:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(a - b);
            break;
          case OpCode.Mul:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(a * b);
            break;
          case OpCode.Div:
            b = this.stack.pop();
            a = this.stack.pop();
            this.stack.push(a / b);
            break;
          case OpCode.Neg:
            a = this.stack.pop();
            this.stack.push(-a);
            break;
          case OpCode.Print:
            this.host.writeLine(this.stack.pop());
            break;
          case OpCode.Jump:
            this.notImplemented('Jump');
            break;
          case OpCode.JumpIfFalse:
            this.notImplemented('JumpIfFalse');
            break;
          case OpCode.Loop:
            this.notImplemented('Loop');
            break;
          case OpCode.Return:
            rv = this.stack.pop();
            // TODO: Clean up the current frame, and only return if we're
            // done with the main program.
            return rv;
        }
      }
    });
  }

  notImplemented(message: string): Value {
    throw new NotImplementedError(message, null);
  }
}
