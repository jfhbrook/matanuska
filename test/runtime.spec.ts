import { describe, test } from 'vitest';
import { t } from './helpers/tap';

import { NotImplementedError } from '../exceptions';
import { OpCode } from '../bytecode/opcodes';

import { chunk } from './helpers/bytecode';
import { testChunk } from './helpers/runtime';

describe('expressions', () => {
  test('255', () => {
    testChunk(
      t,
      chunk({
        constants: [255],
        code: [OpCode.Constant, 0, OpCode.Return],
        lines: [-1, -1, -1],
      }),
    );
  });

  test('123.456', () => {
    testChunk(
      t,
      chunk({
        constants: [123.456],
        code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
    );
  });

  for (const bool of [true, false]) {
    test(`${bool}`, () => {
      testChunk(
        t,
        chunk({
          constants: [bool],
          code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1, -1],
        }),
      );
    });
  }

  test('nil', () => {
    testChunk(
      t,
      chunk({
        constants: [],
        code: [OpCode.Nil, OpCode.Pop, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
    );
  });

  test('"hello world"', () => {
    testChunk(
      t,
      chunk({
        constants: ['hello world'],
        code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
    );
  });

  test('(1)', () => {
    testChunk(
      t,
      chunk({
        constants: [1],
        code: [OpCode.Constant, 0, OpCode.Pop, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1],
      }),
    );
  });

  test('1 + 1', () => {
    testChunk(
      t,
      chunk({
        constants: [1, 1],
        code: [
          OpCode.Constant,
          0,
          OpCode.Constant,
          1,
          OpCode.Add,
          OpCode.Pop,
          OpCode.Nil,
          OpCode.Return,
        ],
        lines: [-1, -1, -1, -1, -1, -1, -1, -1],
      }),
    );
  });

  test('-1', () => {
    testChunk(
      t,
      chunk({
        constants: [1],
        code: [
          OpCode.Constant,
          0,
          OpCode.Neg,
          OpCode.Pop,
          OpCode.Nil,
          OpCode.Return,
        ],
        lines: [-1, -1, -1, -1],
      }),
    );
  });

  test('1 : 1', () => {
    testChunk(
      t,
      chunk({
        constants: [1, 1],
        code: [
          OpCode.Constant,
          0,
          OpCode.Pop,
          OpCode.Constant,
          1,
          OpCode.Return,
        ],
        lines: [-1, -1, -1, -1, -1, -1],
      }),
    );
  });
});

describe('print', () => {
  test('print 255', () => {
    testChunk(
      t,
      chunk({
        constants: [255],
        code: [OpCode.Constant, 0, OpCode.Print, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
    );
  });

  test('print 123.456', () => {
    testChunk(
      t,
      chunk({
        constants: [123.456],
        code: [OpCode.Constant, 0, OpCode.Print, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1, -1],
      }),
    );
  });

  for (const bool of [true, false]) {
    test(`print ${bool}`, () => {
      testChunk(
        t,
        chunk({
          constants: [bool],
          code: [OpCode.Constant, 0, OpCode.Print, OpCode.Nil, OpCode.Return],
          lines: [-1, -1, -1, -1, -1],
        }),
      );
    });
  }

  test('print nil', () => {
    testChunk(
      t,
      chunk({
        constants: [],
        code: [OpCode.Nil, OpCode.Print, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
    );
  });

  test('print "hello world"', () => {
    testChunk(
      t,
      chunk({
        constants: ['hello world'],
        code: [OpCode.Constant, 0, OpCode.Print, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
    );
  });

  test('print (1)', () => {
    testChunk(
      t,
      chunk({
        constants: [1],
        code: [OpCode.Constant, 0, OpCode.Print, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
    );
  });

  test('print 1 + 1', () => {
    testChunk(
      t,
      chunk({
        constants: [1, 1],
        code: [
          OpCode.Constant,
          0,
          OpCode.Constant,
          1,
          OpCode.Add,
          OpCode.Print,
          OpCode.Nil,
          OpCode.Return,
        ],
        lines: [-1, -1, -1, -1, -1, -1, -1],
      }),
    );
  });
});

describe('exit', () => {
  test('exit 1', () => {
    testChunk(
      t,
      chunk({
        constants: [1],
        code: [OpCode.Constant, 0, OpCode.Exit, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
      {
        exitCode: 1,
      },
    );
  });

  test('exit 100', () => {
    testChunk(
      t,
      chunk({
        constants: [100],
        code: [OpCode.Constant, 0, OpCode.Exit, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
      {
        exitCode: 100,
      },
    );
  });

  test('exit "banana"', () => {
    testChunk(
      t,
      chunk({
        constants: ['banana'],
        code: [OpCode.Constant, 0, OpCode.Exit, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
      {
        exitCode: 1,
      },
    );
  });

  test('exit 123.456', () => {
    testChunk(
      t,
      chunk({
        constants: [123.456],
        code: [OpCode.Constant, 0, OpCode.Exit, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1, -1],
      }),
      {
        exitCode: 123,
      },
    );
  });

  test('exit true', () => {
    testChunk(
      t,
      chunk({
        constants: [true],
        code: [OpCode.Constant, 0, OpCode.Exit, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1, -1],
      }),
      {
        exitCode: 1,
      },
    );
  });

  test('exit false', () => {
    testChunk(
      t,
      chunk({
        constants: [false],
        code: [OpCode.Constant, 0, OpCode.Exit, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1, -1],
      }),
      {
        exitCode: 0,
      },
    );
  });

  test('exit nil', () => {
    testChunk(
      t,
      chunk({
        constants: [],
        code: [OpCode.Nil, OpCode.Exit, OpCode.Nil, OpCode.Return],
        lines: [-1, -1, -1, -1, -1],
      }),
      {
        exitCode: 0,
      },
    );
  });

  test('exit 1 + 1', () => {
    testChunk(
      t,
      chunk({
        constants: [1, 1],
        code: [
          OpCode.Constant,
          0,
          OpCode.Constant,
          1,
          OpCode.Add,
          OpCode.Exit,
          OpCode.Nil,
          OpCode.Return,
        ],
        lines: [-1, -1, -1, -1, -1, -1, -1],
      }),
      {
        exitCode: 2,
      },
    );
  });
});

test('simple program', () => {
  testChunk(
    t,
    chunk({
      constants: ['hello world', 'goodbye'],
      code: [
        OpCode.Constant,
        0,
        OpCode.Print,
        OpCode.Constant,
        1,
        OpCode.Print,
        OpCode.Nil,
        OpCode.Return,
      ],
      lines: [100, 100, 100, 200, 200, 200, 200, 200],
    }),
  );
});

test('something not implemented', () => {
  testChunk(
    t,
    chunk({
      constants: [],
      code: [OpCode.Loop, 0xff, 0xff],
      lines: [100],
    }),
    {
      throws: NotImplementedError,
    },
  );
});
