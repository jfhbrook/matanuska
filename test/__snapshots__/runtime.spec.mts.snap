// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`exit > exit "banana" 1`] = `""`;

exports[`exit > exit "banana" 2`] = `""`;

exports[`exit > exit 1 + 1 1`] = `""`;

exports[`exit > exit 1 + 1 2`] = `""`;

exports[`exit > exit 1 1`] = `""`;

exports[`exit > exit 1 2`] = `""`;

exports[`exit > exit 100 1`] = `""`;

exports[`exit > exit 100 2`] = `""`;

exports[`exit > exit 123.456 1`] = `""`;

exports[`exit > exit 123.456 2`] = `""`;

exports[`exit > exit false 1`] = `""`;

exports[`exit > exit false 2`] = `""`;

exports[`exit > exit nil 1`] = `""`;

exports[`exit > exit nil 2`] = `""`;

exports[`exit > exit true 1`] = `""`;

exports[`exit > exit true 2`] = `""`;

exports[`expressions > "hello world" 1`] = `""`;

exports[`expressions > "hello world" 2`] = `""`;

exports[`expressions > (1) 1`] = `""`;

exports[`expressions > (1) 2`] = `""`;

exports[`expressions > -1 1`] = `""`;

exports[`expressions > -1 2`] = `""`;

exports[`expressions > 1 + 1 1`] = `""`;

exports[`expressions > 1 + 1 2`] = `""`;

exports[`expressions > 1 : 1 1`] = `""`;

exports[`expressions > 1 : 1 2`] = `""`;

exports[`expressions > 123.456 1`] = `""`;

exports[`expressions > 123.456 2`] = `""`;

exports[`expressions > 255 1`] = `""`;

exports[`expressions > 255 2`] = `""`;

exports[`expressions > false 1`] = `""`;

exports[`expressions > false 2`] = `""`;

exports[`expressions > nil 1`] = `""`;

exports[`expressions > nil 2`] = `""`;

exports[`expressions > true 1`] = `""`;

exports[`expressions > true 2`] = `""`;

exports[`print > print "hello world" 1`] = `
"hello world
"
`;

exports[`print > print "hello world" 2`] = `""`;

exports[`print > print (1) 1`] = `
"1
"
`;

exports[`print > print (1) 2`] = `""`;

exports[`print > print 1 + 1 1`] = `
"2
"
`;

exports[`print > print 1 + 1 2`] = `""`;

exports[`print > print 123.456 1`] = `
"123.456
"
`;

exports[`print > print 123.456 2`] = `""`;

exports[`print > print 255 1`] = `
"255
"
`;

exports[`print > print 255 2`] = `""`;

exports[`print > print false 1`] = `
"false
"
`;

exports[`print > print false 2`] = `""`;

exports[`print > print nil 1`] = `
"nil
"
`;

exports[`print > print nil 2`] = `""`;

exports[`print > print true 1`] = `
"true
"
`;

exports[`print > print true 2`] = `""`;

exports[`simple program 1`] = `
"hello world
goodbye
"
`;

exports[`simple program 2`] = `""`;

exports[`something not implemented 1`] = `
"Traceback:
  File '<unknown>', line 100
NotImplementedError: Loop"
`;
